### ISSUE TITLE
add 'list' data structure support or code learning path

### ISSUE THREAD

**Issue by YouhuaLi** [sorear/metamath-turing-machines#6]
Hello, this is not a real issue but a general question. 

Do you have any plan to add 'list' support in NQL recently? If not, I am also interested to add the support myself.  I checked http://us.metamath.org  but seems there will be years of reading and learning.

So can you give the shortest learning path for a person to develop and extend NQL, assuming this person has good knowledge of python, Turing machine, original lacnoic language, but lack of knowledge of meta math?

Thanks a lot.
Youhua Li

**Comment by sorear** [comment]
If you want Lisp-style lists, copy pair and unpair from zf.nql into your own file.  It doesn't depend on anything else in the file.  Original Laconic is also not terribly relevant except as inspiration.

The NQL model (register machine + relative jumps + binary decision diagram) is pushed to what I think is its limit; I don't really see any need to do new development on it.  Are you trying to demonstrate something specific?

(Did this get in the news again?)

**Comment by YouhuaLi** [comment]
I found the pair/unpair procs almost solve my problems. And yes, I am trying to demonstrate something specific.  I want to prove something when a Turing machine halts, while what I see now are something when Turing machines won't halt (at least expected). I will get back to you on this soon.

(And yes, I got this in the news again on a Chinese media recently.)

**Comment by YouhuaLi** [comment]
all right,  I committed an NQL file: a [turing machine](https://github.com/YouhuaLi/metamath-turing-machines/blob/master/ackermann.nql) can computer Ackermann function.  

My idea is to use this way to improve the lower bound of Rado's [sigma function](https://googology.wikia.org/wiki/Busy_beaver_function). If a two-symbol, n-state turning machine can compute a very large number x, that means S(n)>x.  

I have done this with Yedidia's [lacnoic](https://github.com/YouhuaLi/parsimony/blob/5b9aacf575ead530f9c2af9065570d96f49415f3/src/laconic/laconic_files/ackermann.lac) language, but it requires a 5000+ states.  That's kind of a prove of S(5000)>A(A(5), A(5)). But lately, I learned that because there is a 1919-state machine that can prove the consistency of ZFC, which means S(1919) is greater than all integers can be defined and computable in ZFC (assume ZFC is consistency).

Then I found NQL is much more efficient in the context of state count.   Now it seems a 500-state nql program can compute Ackermann function before optimization which is a quite nice result.  I have last two questions:

1) is there a better way to test if a stack is empty?  Now I use "0" to indicate empty stack and I have to increase value by 1 before push a variable into a stack and decease 1 after a pop operation. I wonder if there is a better way for this.

2) is there a way to print variable value when running nql program for debugging purposes? I know --print-subs can help but I am not able to read it well.  A print function would be very helpful. 

Thanks,
Youhua
